### DISCUSSION TITLE
Tensor datatypes

### DISCUSSION THREAD

**Discussion by soerenwolfers** [duckdb/duckdb#13736] [Ideas]
It would often be convenient to have tensor data types instead of nested lists or arrays.

To make everything below as explicit as possible, I am writing all examples using fixed-shape tensors
```
T[length1, ..., lengthN]
```
which generalize the `ARRAY` types `T[length1]` to rank `N >= 1`.

In practice, because it's rare that one has data where more than the rank is known a-priori, I believe it would be often more useful to have generalizations of `LIST` types, either via fixed-rank-but-arbitrary-lengths tensors (called multi-dimensional arrays in most languages) 
```
T[] -- List = Tensor of rank 1 with arbitrary length, already exists
T[,]  -- Two-dimensional array = Tensor of rank 2 with arbitrary lengths
T[,,] -- Three-dimensional array = Tensor rank 3 with arbitrary lengths, etc.
-- e.g., one row with multiple columns for different ranks between 1 and 3 via
VALUES 
    ([1, 2]::INT[], [[1, 2], [3, 4]]::INT[,], [[[1], [[2]], [[3]]]::INT[,,])
```
or via arbitrary-shape tensors (the equivalent of numpy's `ndarray` and PG's multi-dimensional array)
```
T[...] -- can hold any rank and any lengths
-- e.g. a single column with entries of ranks between 0 and 3 via 
VALUES
  (1::INT[...]), 
  ([1, 2]::INT[...]),
  ([[1, 2], [3, 4]]::INT[...]), 
  ([[[1], [[2]], [[3]]]::INT[...]),
```
All examples below apply to both of these these with some obvious modifications. 

I imagine some might be against tensors in a database, with the argument that the point of having a database is to have a long and tidy data format and use SQL syntax to reshuffle data. My counterarguments to this hypothetical objection, besides the fact that PostgreSQL already does it, would be 
- many users I know like SQL to do the hard bits (like handling timestamps, missing data, filtering, and aggregation) but prefer doing the simple bits, e.g., reshaping/transposing/multiplying "square arrays", using simple syntax like numpy's rather than forcing a unnested format augmented by indices to be kept track of in SQL.
- while everything that tensors can do can already be done in SQL, it's often way too slow in SQL (e.g. matrix mutliplications or reshaping)
- the `LIST`/`ARRAY` types exist for broadly the same reasons

(All examples below use zero-indexing because I accidentally started writing this proposal as such and changing it now would just introduce errors.)

**Construction:** 
```sql 
SELECT [1, 2, 3]::INT[3] UNION SELECT [4, 5, 6]::INT[3] --- Current ARRAY column of type INT[3] holds tensors of rank 1 and length 3
SELECT [[1, 2, 3], [4, 5, 6]]::INT[2, 3] UNION SELECT [[5, 6, 7], [8, 9, 10]]::INT[2, 3] --- New TENSOR column of type INT[2, 3] holds tensors of rank 2 and shape [2, 3]
```

**Access:** 
```sql 
SELECT [[1, 2, 3], [4, 5, 6]]::INT[2, 3][0, 1] == 2 --- Row major layout ensures `a[0][1] == (a::INT[2, 3])[0, 1]` for a nested list `a` and ensures that `len(a) == shape(a::INT[2, 3])[0]`
SELECT [[1, 2, 3], [4, 5, 6]]::INT[2, 3][0] --- ERROR: No implicit cast from INT[2, 3] to INT[] or INT[6] / no implicit padding of indices.
SELECT [[1, 2, 3], [4, 5, 6]]::INT[2, 3][0, 1..] == [2, 3]::INT[2] --- slicing
```

**Conversion**

```sql 
SELECT [5, 6, 7, 8, 9, 10]::INT[2, 3] == [[5, 6, 7], [8, 9, 10]]::INT[2, 3] --- Explicit conversion from INT[]/INT[z] and INT[][] to INT[x, y] when lengths match, by reinterpreting the row-major data in contiguous memory with different shape arguments.
SELECT [[5, 6], [7, 8]]::INT[2, 2]::INT[] --- Explicit conversion from INT[x, y] to INT[](/INT[z] when lengths match), by reinterpreting the row-major data in contiguous memory with different shape arguments
SELECT [[5, 6], [7, 8]]::INT[3, 2]::INT[2, 3] --- Explicit conversion from INT[x, y] to INT[a, b] when lengths match, by reinterpreting the row-major data in contiguous memory with different shape arguments
SELECT [5, 6, 7, 8]::INT[]::INT[2, 1] --- Errors when lengths don't match
SELECT [5, 6, 7, 8]::INT[]::INT[2, 3] --- No automatic padding

SELECT [[1, 2], [3, 4]]::INT[2, 2] UNION SELECT [[5], [6]]::INT[2, 1] --- ERROR. No implicit conversion between INT[2, 2] and INT[2, 1]
SELECT [[1, 2], [3, 4]]::INT[2, 2] UNION SELECT [5, 6, 7, 8]::INT[4] --- ERROR: No implicit conversion between INT[2, 2] and INT[4]
SELECT [[1, 2], [3, 4]]::INT[2, 2] UNION SELECT [5, 6, 7, 8]::INT[4, 1] --- ERROR: No implicit conversion between INT[2, 2] and INT[4, 1]

SELECT [[1, 2], [3, 4]]::INT[2, 2] UNION SELECT [5, 6]::INT[4] --- ERROR: No implicit conversion between INT[2, 2] and INT[4]
SELECT [[1, 2], [3, 4]]::INT[2, 2] UNION SELECT [[5], [6]]::INT[2, 1] --- ERROR: Cannot form union between INT[2, 2] and INT[2, 1]
```

**Operations**

To start with, the 1% of most common duckdb.list and numpy.ndarray operations would already be great. More can be added later.

 For example:

```sql
transform([[1, 2], [3, 4]]::INT[2, 2], x -> 2 * x) == [[2, 4], [6, 8]]::INT[2, 2] --- same as array_transform
transpose([[[1], [2]], [[3], [4]], [[5], [6]]]::INT[3, 2, 1]) == [[[1, 3, 5], [2, 4, 6]]]::INT[1, 2, 3] --- same as numpy.transpose. shape(x) == reverse(shape(transpose(x)))
[[1, 2], [3, 4]]::INT[2, 2] @ [1, 2]::INT[2] == [5, 11]::INT[2] --- matrix vector multiplication. Only available on numeric types. 
[[1, 2], [3, 4]]::INT[2, 2] @ [1, 2]::INT[] == [5, 11]::INT[2] --- Implicit conversion of second argument to INT[2]
[[1, 2], [3, 4]]::INT[2, 2] @ [[1], [2]]::INT[2, 1] == [[5], [11]]::INT[2, 1] --- matrix matrix multiplication 
sum([[1, 2], [3, 4]]::INT[2, 2], 0) == [4, 6]::INT[2]  --- sum across zeroth axis. only available on numeric types
```

**Comment by Maxxen** [comment]
Can you not use nested lists/arrays for this? Im planning to add more numpy-like broadcast-style functions to the array type in general in the future.

**Reply by soerenwolfers** [reply]
Besides nested lists not guaranteeing "squareness", the problem is that nested lists don't (often cannot and sometimes should not) offer functionality that's inherent to the tensor interpretation: no reshapes, no transposes, no slicing, no matrix multiplications, no sum-across-dimension, no apply-transform-to-all-elements, no conversion from list to list-of-lists

**Reply by soerenwolfers** [reply]
some evidence that tensors are popular and needed are PostgreSQL's multi-dimensional array, the hugely popular numpy.ndarray, the xarray library, and the fact that Microsoft recently added tensors to .NET8 (https://devblogs.microsoft.com/dotnet/introducing-tensor-for-multi-dimensional-machine-learning-and-ai-data/)

**Reply by soerenwolfers** [reply]
that being said, if you think adding tensor types would be much more effort than just writing the functions (that throw on mishaped data) I'd of course happily accept the notational inconvenience and be grateful for the functionality

**Reply by soerenwolfers** [reply]
Although, for acceptably fast tensor operations (especially matrix multiplications) you really need memory locality so nested lists will likely not be good enough.

**Comment by rustyconover** [comment]
@soerenwolfers Checkout PR https://github.com/duckdb/duckdb/pull/22082
